tests/tests.cpp

```diff
@@ -0,0 +1,398 @@
+    }
+#endif
+
+    TEST_F(CompletenessTest, Attestation) {
+        // Currently the unittests only support building for the host, so we
+        // can't generate evidence. Instead we verify pre-generated evidence
+
+        // Load nonce, public key, enclave signing key, and evidence from file
+        std::vector<uint8_t> nonce;
+        std::vector<uint8_t> pub_key;
+        std::vector<uint8_t> signing_key;
+        std::vector<uint8_t> evidence;
+        load_file(std::string("../../tests/data/nonce"), nonce);
+        load_file(std::string("../../tests/data/report_key.pub"), pub_key);
+        load_file(std::string("../../tests/data/signing_key.pub"), signing_key);
+        load_file(std::string("../../tests/data/evidence"), evidence);
+
+        // Verify the evidence
+        ASSERT_EQ(0, attestation->AttestEvidence(
+                         &sgx_remote_uuid, signing_key.data(), evidence.data(),
+                         pub_key.data(), nonce.data(), signing_key.size() + 1,
+                         evidence.size(), pub_key.size()));
+    }
+
+    // ###########################
+    // ##### SOUNDNESS TESTS #####
+    // ###########################
+
+    // Dummy class to create a new test suite
+    class SoundnessTest : public UtilsTest {};
+
+    TEST_F(SoundnessTest, SymEnc) {
+        // Encrypt the message
+        std::vector<uint8_t> ct(crypto.SymEncSize(msg.size()));
+        ASSERT_EQ(0, crypto.SymEnc(sym_key, msg.data(),
+                                   reinterpret_cast<const uint8_t*>("aad"),
+                                   ct.data(), msg.size(), strlen("aad")));
+
+        // We don't care about the value of the decryptions (they all should
+        // fail) so we can use the same plaintext vector for each decryption
+        std::vector<uint8_t> pt(crypto.SymDecSize(ct.size()));
+
+        // 1) Make random pertubations in the ciphertexts
+        for (int i = 0; i < 100; i++) {
+            std::vector<uint8_t> perturbed_ct(ct);
+            perturb(perturbed_ct);
+            ASSERT_NE(0, crypto.SymDec(sym_key, perturbed_ct.data(),
+                                       reinterpret_cast<const uint8_t*>("aad"),
+                                       pt.data(), perturbed_ct.size(),
+                                       strlen("aad")))
+                << "Perturbed ciphertext decrypted successfully";
+        }
+
+        // 2) Change the IV
+        uint8_t new_iv[CIPHER_IV_SIZE];
+        crypto.RandGen(new_iv, CIPHER_IV_SIZE);
+
+        std::vector<uint8_t> perturbed_ct(ct);
+        for (int i = 0; i < CIPHER_IV_SIZE; i++) {
+            perturbed_ct[i] = new_iv[i];
+        }
+        ASSERT_NE(0,
+                  crypto.SymDec(sym_key, perturbed_ct.data(),
+                                reinterpret_cast<const uint8_t*>("aad"),
+                                pt.data(), perturbed_ct.size(), strlen("aad")))
+            << "Ciphertext with invalid IV decrypted successfully";
```

Comment:
So the current output on a failed test is something like:
```
6: /home/ryan/mc2-utils/tests/tests.cpp:201: Failure                                                                                                    
6: Expected equality of these values:                                                                
6:   0                                                                                                                                                            
6:   crypto.SymDec(sym_key, perturbed_ct.data(), reinterpret_cast<const uint8_t*>("aad"), pt.data(), perturbed_ct.size(), strlen("aad"))
6:     Which is: -18                                                                                                                                                      
6: Perturbed ciphertext decrypted successfully                                                  
6: 2021-06-30 22:58:27 - HOST - info - Successfully freed relevant cryptography contexts.                                                                                 
6: [  FAILED  ] SoundnessTest.SymEnc (79 ms)                                                                                  
6: [----------] 1 test from SoundnessTest (79 ms total)
```

Do you think it would still be clearer to add the ERROR as well?

Reply:
Ah I see. This is probably fine then. When I see "successfully" I usually think everything ran ok, but the Failure message should clarify this.